File: _skels/go-ddd-skel/internal/orders/service.go
// Service-layer logic for /api/orders. The service orchestrates
// the orders aggregate (Order + OrderLine + OrderAddress) and the
// cross-resource catalog lookup that snapshots the unit price into
// each line.
package orders

import (
	"context"
	"errors"
	"fmt"
	"time"

	"github.com/example/go-ddd-skel/internal/shared"
)

// Service is the entry point for the orders aggregate. Holds an
// orders.Repository plus a CatalogReader so cross-resource
// orchestration (line creation copies catalog price) is just a
// pure-Go method call.
type Service struct {
	orders  Repository
	catalog CatalogReader
}

// NewService builds an orders Service.
func NewService(repo Repository, cat CatalogReader) *Service {
	return &Service{orders: repo, catalog: cat}
}

// CreateOrder starts a fresh draft order owned by userID.
func (s *Service) CreateOrder(ctx context.Context, userID uint) (Order, error) {
	order := Order{UserID: userID, Status: "draft"}
	if err := s.orders.CreateOrder(ctx, &order); err != nil {
		return Order{}, err
	}
	return order, nil
}

// ListOrders returns every order owned by userID.
func (s *Service) ListOrders(ctx context.Context, userID uint) ([]Order, error) {
	return s.orders.ListOrdersForUser(ctx, userID)
}

// GetDetail fetches an order plus its lines + address. Returns
// shared.ErrNotFound when the order is missing or owned by someone
// else.
func (s *Service) GetDetail(ctx context.Context, orderID, userID uint) (Detail, error) {
	order, err := s.fetchUserOrder(ctx, orderID, userID)
	if err != nil {
		return Detail{}, err
	}
	return s.buildDetail(ctx, order)
}

// AddLine appends a new line to a draft order, snapshotting the
// catalog item's current price.
func (s *Service) AddLine(ctx context.Context, orderID, userID uint, dto AddLineDTO) (OrderLine, error) {
	order, err := s.fetchUserOrder(ctx, orderID, userID)
	if err != nil {
		return OrderLine{}, err
	}
	if order.Status != "draft" {
		return OrderLine{}, fmt.Errorf("%w: can only add lines to draft orders", shared.ErrValidation)
	}
	if dto.CatalogItemID == 0 {
		return OrderLine{}, fmt.Errorf("%w: catalog_item_id is required", shared.ErrValidation)
	}
	qty := dto.Quantity
	if qty < 1 {
		qty = 1
	}
	catItem, err := s.catalog.Get(ctx, dto.CatalogItemID)
	if err != nil {
		return OrderLine{}, err
	}
	line := OrderLine{
		OrderID:       orderID,
		CatalogItemID: dto.CatalogItemID,
		Quantity:      qty,
		UnitPrice:     catItem.Price,
	}
	if err := s.orders.CreateLine(ctx, &line); err != nil {
		return OrderLine{}, err
	}
	return line, nil
}

// DeleteLine removes a line from a draft order. Returns
// shared.ErrNotFound when the order or line does not exist (or is
// owned by someone else).
func (s *Service) DeleteLine(ctx context.Context, orderID, userID, lineID uint) error {
	order, err := s.fetchUserOrder(ctx, orderID, userID)
	if err != nil {
		return err
	}
	if order.Status != "draft" {
		return fmt.Errorf("%w: can only remove lines from draft orders", shared.ErrValidation)
	}
	rows, err := s.orders.DeleteLine(ctx, orderID, lineID)
	if err != nil {
		return err
	}
	if rows == 0 {
		return fmt.Errorf("%w: order line %d", shared.ErrNotFound, lineID)
	}
	return nil
}

// UpsertAddress creates or replaces the delivery address for an
// order owned by userID.
func (s *Service) UpsertAddress(ctx context.Context, orderID, userID uint, dto AddressDTO) (OrderAddress, error) {
	if _, err := s.fetchUserOrder(ctx, orderID, userID); err != nil {
		return OrderAddress{}, err
	}
	addr, err := s.orders.GetAddressForOrder(ctx, orderID)
	switch {
	case err == nil:
		addr.Street = dto.Street
		addr.City = dto.City
		addr.ZipCode = dto.ZipCode
		addr.Phone = dto.Phone
		addr.Notes = dto.Notes
		if err := s.orders.SaveAddress(ctx, &addr); err != nil {
			return OrderAddress{}, err
		}
		return addr, nil
	case errors.Is(err, shared.ErrNotFound):
		addr = OrderAddress{
			OrderID: orderID,
			Street:  dto.Street,
			City:    dto.City,
			ZipCode: dto.ZipCode,
			Phone:   dto.Phone,
			Notes:   dto.Notes,
		}
		if err := s.orders.SaveAddress(ctx, &addr); err != nil {
			return OrderAddress{}, err
		}
		return addr, nil
	default:
		return OrderAddress{}, err
	}
}

// Submit transitions a draft order to pending. Requires at least
// one line.
func (s *Service) Submit(ctx context.Context, orderID, userID uint) (Detail, error) {
	order, err := s.fetchUserOrder(ctx, orderID, userID)
	if err != nil {
		return Detail{}, err
	}
	if order.Status != "draft" {
		return Detail{}, fmt.Errorf("%w: only draft orders can be submitted", shared.ErrValidation)
	}
	lineCount, err := s.orders.CountLinesForOrder(ctx, orderID)
	if err != nil {
		return Detail{}, err
	}
	if lineCount == 0 {
		return Detail{}, fmt.Errorf("%w: cannot submit an order with no lines", shared.ErrValidation)
	}
	now := time.Now().UTC()
	order.Status = "pending"
	order.SubmittedAt = &now
	if err := s.orders.SaveOrder(ctx, &order); err != nil {
		return Detail{}, err
	}
	return s.buildDetail(ctx, order)
}

// Approve transitions a pending order to approved.
func (s *Service) Approve(ctx context.Context, orderID, userID uint, dto ApproveDTO) (Detail, error) {
	order, err := s.fetchUserOrder(ctx, orderID, userID)
	if err != nil {
		return Detail{}, err
	}
	if order.Status != "pending" {
		return Detail{}, fmt.Errorf("%w: only submitted orders can be approved", shared.ErrValidation)
	}
	order.Status = "approved"
	order.WaitMinutes = dto.WaitMinutes
	order.Feedback = dto.Feedback
	if err := s.orders.SaveOrder(ctx, &order); err != nil {
		return Detail{}, err
	}
	return s.buildDetail(ctx, order)
}

// Reject transitions a pending order to rejected.
func (s *Service) Reject(ctx context.Context, orderID, userID uint, dto RejectDTO) (Detail, error) {
	order, err := s.fetchUserOrder(ctx, orderID, userID)
	if err != nil {
		return Detail{}, err
	}
	if order.Status != "pending" {
		return Detail{}, fmt.Errorf("%w: only submitted orders can be rejected", shared.ErrValidation)
	}
	order.Status = "rejected"
	order.Feedback = dto.Feedback
	if err := s.orders.SaveOrder(ctx, &order); err != nil {
		return Detail{}, err
	}
	return s.buildDetail(ctx, order)
}

// fetchUserOrder loads an order by id and verifies it belongs to
// the given user. Returns shared.ErrNotFound when not found OR not
// owned (so callers map both to 404, matching the prior contract).
func (s *Service) fetchUserOrder(ctx context.Context, orderID, userID uint) (Order, error) {
	o, err := s.orders.GetOrder(ctx, orderID)
	if err != nil {
		return o, err
	}
	if o.UserID != userID {
		return o, fmt.Errorf("%w: order %d not owned by user %d", shared.ErrNotFound, orderID, userID)
	}
	return o, nil
}

// buildDetail composes lines + address around an Order to produce
// the response shape every Detail-returning endpoint shares.
func (s *Service) buildDetail(ctx context.Context, order Order) (Detail, error) {
	lines, err := s.orders.ListLinesForOrder(ctx, order.ID)
	if err != nil {
		return Detail{}, err
	}
	if lines == nil {
		lines = []OrderLine{}
	}
	var address *OrderAddress
	addr, addrErr := s.orders.GetAddressForOrder(ctx, order.ID)
	switch {
	case addrErr == nil:
		address = &addr
	case errors.Is(addrErr, shared.ErrNotFound):
		address = nil
	default:
		return Detail{}, addrErr
	}
	return Detail{Order: order, Lines: lines, Address: address}, nil
}

File: _skels/rust-axum-ddd-skel/src/orders/service.rs
//! Service-layer logic for `/api/orders`. Orchestrates the orders
//! aggregate (Order + OrderLine + OrderAddress) and the cross-
//! resource catalog lookup that snapshots the unit price into each
//! line.

use std::sync::Arc;

use chrono::Utc;
use serde::Serialize;

use crate::catalog::service::CatalogService;
use crate::orders::repository::{
    AddLine, NewAddress, Order, OrderAddress, OrderLine, OrderRepository, OrderStatusUpdate,
};
use crate::shared::DomainError;

#[derive(Debug, Clone)]
pub struct AddLineDTO {
    pub catalog_item_id: i32,
    pub quantity: i32,
}

#[derive(Debug, Clone)]
pub struct AddressDTO {
    pub street: String,
    pub city: String,
    pub zip_code: String,
    pub phone: Option<String>,
    pub notes: Option<String>,
}

#[derive(Debug, Clone)]
pub struct ApproveDTO {
    pub wait_minutes: Option<i32>,
    pub feedback: Option<String>,
}

#[derive(Debug, Clone)]
pub struct RejectDTO {
    pub feedback: Option<String>,
}

/// Response DTO for `/api/orders/:id` and every endpoint that
/// returns the fully-loaded order. The shape matches the prior
/// handler-monolith JSON byte-for-byte.
#[derive(Debug, Clone, Serialize)]
pub struct OrderDetail {
    pub id: i32,
    pub user_id: i32,
    pub status: String,
    pub created_at: chrono::DateTime<Utc>,
    pub submitted_at: Option<chrono::DateTime<Utc>>,
    pub wait_minutes: Option<i32>,
    pub feedback: Option<String>,
    pub lines: Vec<OrderLine>,
    pub address: Option<OrderAddress>,
}

impl OrderDetail {
    fn from_parts(order: Order, lines: Vec<OrderLine>, address: Option<OrderAddress>) -> Self {
        Self {
            id: order.id,
            user_id: order.user_id,
            status: order.status,
            created_at: order.created_at,
            submitted_at: order.submitted_at,
            wait_minutes: order.wait_minutes,
            feedback: order.feedback,
            lines,
            address,
        }
    }
}

#[derive(Clone)]
pub struct OrdersService {
    repo: Arc<dyn OrderRepository>,
    catalog: CatalogService,
}

impl OrdersService {
    pub fn new(repo: Arc<dyn OrderRepository>, catalog: CatalogService) -> Self {
        Self { repo, catalog }
    }

    pub async fn create_order(&self, user_id: i32) -> Result<Order, DomainError> {
        self.repo.create_order(user_id).await
    }

    pub async fn list_orders(&self, user_id: i32) -> Result<Vec<Order>, DomainError> {
        self.repo.list_orders_for_user(user_id).await
    }

    pub async fn get_detail(
        &self,
        order_id: i32,
        user_id: i32,
    ) -> Result<OrderDetail, DomainError> {
        let order = self.fetch_user_order(order_id, user_id).await?;
        self.build_detail(order).await
    }

    pub async fn add_line(
        &self,
        order_id: i32,
        user_id: i32,
        dto: AddLineDTO,
    ) -> Result<OrderLine, DomainError> {
        let order = self.fetch_user_order(order_id, user_id).await?;
        if order.status != "draft" {
            return Err(DomainError::Validation(
                "can only add lines to a draft order".to_string(),
            ));
        }
        if dto.quantity < 1 {
            return Err(DomainError::Validation(
                "quantity must be at least 1".to_string(),
            ));
        }
        let cat = self.catalog.get(dto.catalog_item_id).await?;
        self.repo
            .create_line(AddLine {
                order_id,
                catalog_item_id: dto.catalog_item_id,
                quantity: dto.quantity,
                unit_price: cat.price,
            })
            .await
    }

    pub async fn delete_line(
        &self,
        order_id: i32,
        user_id: i32,
        line_id: i32,
    ) -> Result<(), DomainError> {
        let order = self.fetch_user_order(order_id, user_id).await?;
        if order.status != "draft" {
            return Err(DomainError::Validation(
                "can only remove lines from a draft order".to_string(),
            ));
        }
        let rows = self.repo.delete_line(order_id, line_id).await?;
        if rows == 0 {
            return Err(DomainError::NotFound(format!(
                "order line {line_id} not found"
            )));
        }
        Ok(())
    }

    pub async fn upsert_address(
        &self,
        order_id: i32,
        user_id: i32,
        dto: AddressDTO,
    ) -> Result<OrderAddress, DomainError> {
        let order = self.fetch_user_order(order_id, user_id).await?;
        if order.status != "draft" {
            return Err(DomainError::Validation(
                "can only set address on a draft order".to_string(),
            ));
        }
        if dto.street.trim().is_empty()
            || dto.city.trim().is_empty()
            || dto.zip_code.trim().is_empty()
        {
            return Err(DomainError::Validation(
                "street, city, and zip_code are required".to_string(),
            ));
        }
        self.repo
            .upsert_address(
                order_id,
                NewAddress {
                    street: dto.street,
                    city: dto.city,
                    zip_code: dto.zip_code,
                    phone: dto.phone.unwrap_or_default(),
                    notes: dto.notes.unwrap_or_default(),
                },
            )
            .await
    }

    pub async fn submit(&self, order_id: i32, user_id: i32) -> Result<Order, DomainError> {
        let order = self.fetch_user_order(order_id, user_id).await?;
        if order.status != "draft" {
            return Err(DomainError::Validation(
                "only draft orders can be submitted".to_string(),
            ));
        }
        let line_count = self.repo.count_lines_for_order(order_id).await?;
        if line_count == 0 {
            return Err(DomainError::Validation(
                "cannot submit an order with no lines".to_string(),
            ));
        }
        self.repo
            .apply_status_update(
                order_id,
                OrderStatusUpdate {
                    status: "pending".to_string(),
                    submitted_at: Some(Some(Utc::now())),
                    wait_minutes: None,
                    feedback: None,
                },
            )
            .await
    }

    pub async fn approve(&self, order_id: i32, dto: ApproveDTO) -> Result<Order, DomainError> {
        let order = self.repo.get_order(order_id).await?;
        if order.status != "pending" {
            return Err(DomainError::Validation(
                "only submitted orders can be approved".to_string(),
            ));
        }
        self.repo
            .apply_status_update(
                order_id,
                OrderStatusUpdate {
                    status: "approved".to_string(),
                    submitted_at: None,
                    wait_minutes: Some(dto.wait_minutes),
                    feedback: Some(dto.feedback),
                },
            )
            .await
    }

    pub async fn reject(&self, order_id: i32, dto: RejectDTO) -> Result<Order, DomainError> {
        let order = self.repo.get_order(order_id).await?;
        if order.status != "pending" {
            return Err(DomainError::Validation(
                "only submitted orders can be rejected".to_string(),
            ));
        }
        self.repo
            .apply_status_update(
                order_id,
                OrderStatusUpdate {
                    status: "rejected".to_string(),
                    submitted_at: None,
                    wait_minutes: None,
                    feedback: Some(dto.feedback),
                },
            )
            .await
    }

    pub async fn delete(&self, order_id: i32, user_id: i32) -> Result<(), DomainError> {
        let order = self.fetch_user_order(order_id, user_id).await?;
        if order.status != "draft" {
            return Err(DomainError::Validation(
                "only draft orders can be deleted".to_string(),
            ));
        }
        // Aggregate cleanup — explicit so the operation works whether
        // or not SQLite's FK cascade is enabled (and matches the
        // prior handler implementation byte-for-byte).
        self.repo.delete_lines_for_order(order_id).await?;
        self.repo.delete_addresses_for_order(order_id).await?;
        self.repo.delete_order(order_id).await
    }

    async fn fetch_user_order(
        &self,
        order_id: i32,
        user_id: i32,
    ) -> Result<Order, DomainError> {
        let order = self.repo.get_order(order_id).await?;
        if order.user_id != user_id {
            return Err(DomainError::NotFound(format!(
                "order {order_id} not found"
            )));
        }
        Ok(order)
    }

    async fn build_detail(&self, order: Order) -> Result<OrderDetail, DomainError> {
        let id = order.id;
        let lines = self.repo.list_lines_for_order(id).await?;
        let address = self.repo.get_address_for_order(id).await?;
        Ok(OrderDetail::from_parts(order, lines, address))
    }
}
